### PR TITLE
optionally specify location when creating worker

### PR DESCRIPTION
**Pull request by emord**
https://docs.google.com/document/d/1Ta9TMnDHxNbYT-oa8_3HbejyK0wsSJXkXD_MbMHZIZY/edit#heading=h.pxre0b5l5zxj

Allows you to optionally set a location when creating a new mobile worker. uses select2

buddy @czue

### CODE REVIEW

File: corehq/apps/users/forms.py [hunk 1]
```diff
@@ -485,18 +485,37 @@ class NewMobileWorkerForm(forms.Form):
...
         super(NewMobileWorkerForm, self).__init__(*args, **kwargs)
-        email_string = u"@{}.commcarehq.org".format(domain)
+        email_string = u"@{}.commcarehq.org".format(project.name)
         max_chars_username = 80 - len(email_string)
 
+        if project.uses_locations:
+            locations = SQLLocation.by_domain(project.name)
+            self.fields['location_id'].choices = [(l.location_id, l.name) for l in locations]
+            location_field = crispy.Field(
+                    'location_id',
```
Comment (dimagimon) [inline corehq/apps/users/forms.py]: E126 continuation line over-indented for hanging indent

File: corehq/apps/users/forms.py [hunk 2]
```diff
@@ -485,18 +485,37 @@ class NewMobileWorkerForm(forms.Form):
...
+                    ng_model='mobileWorker.location_id',
+                )
```
Comment (dimagimon) [inline corehq/apps/users/forms.py]: E121 continuation line under-indented for hanging indent

File: corehq/apps/users/forms.py [hunk 3]
```diff
@@ -485,18 +485,37 @@ class NewMobileWorkerForm(forms.Form):
...
-    def __init__(self, domain, *args, **kwargs):
+    def __init__(self, project, *args, **kwargs):
         super(NewMobileWorkerForm, self).__init__(*args, **kwargs)
-        email_string = u"@{}.commcarehq.org".format(domain)
+        email_string = u"@{}.commcarehq.org".format(project.name)
         max_chars_username = 80 - len(email_string)
 
+        if project.uses_locations:
+            locations = SQLLocation.by_domain(project.name)
+            self.fields['location_id'].choices = [(l.location_id, l.name) for l in locations]
```
Comment (czue) [inline corehq/apps/users/forms.py]: nit: it would likely be more performant to use `values_list` (less IO, less python wrapping)

**Comment by czue** [comment]
do you have any performance concerns about loading every available location (id, name) in the domain into memory and into the front end? Could be wrong, but I thought most of the `select2` location selectors used a server API to populate the choices.

File: corehq/apps/users/static/users/js/mobile_workers.ng.js [hunk 4]
```diff
@@ -94,6 +96,20 @@
...
+        $scope.availableLocations = [];
+
+        $scope.searchLocations = function (query) {
+            var reqStr = location_url + "?name=" + query;
+            $http.get(reqStr).then(
+                function (response) {
+                    $scope.availableLocations = response.data;
+                },
+                function () {
+                    console.log('Error retrieving locations');
```
Comment (dimagimon) [inline corehq/apps/users/static/users/js/mobile_workers.ng.js]: Unexpected console statement. (no-console)

File: corehq/apps/users/forms.py [hunk 5]
```diff
@@ -18,7 +18,7 @@
 from corehq import toggles
 from corehq.apps.domain.forms import EditBillingAccountInfoForm
 from corehq.apps.domain.models import Domain
-from corehq.apps.locations.models import Location
+from corehq.apps.locations.models import Location, SQLLocation
```
Comment (dimagimon) [inline corehq/apps/users/forms.py]: F401 'SQLLocation' imported but unused

File: corehq/apps/users/forms.py [hunk 6]
```diff
@@ -615,6 +634,28 @@ def __init__(self, *args, **kwargs):
...
+    """
+
+    def __init__(self, attrs=None):
+        super(AngularLocationSelectWidget, self).__init__(attrs)
+
+    def render(self, name, value, attrs=None):
+        return """
+          <ui-select  ng-model="mobileWorker.location_id" theme="select2" style="width: 300px;">
+            <ui-select-match placeholder="Select location...">{{$select.selected.name}}</ui-select-match>
+            <ui-select-choices refresh="searchLocations($select.search)" refresh-delay="0" repeat="location.id as location in availableLocations | filter:$select.search">
```
Comment (dimagimon) [inline corehq/apps/users/forms.py]: E501 line too long (170 > 115 characters)

**Comment by emord** [comment]
@czue  I couldn't use the existing widget because angular doesn't play nicely with hidden inputs. https://github.com/dimagi/commcare-hq/pull/12631/commits/7968f37b6c37896b3cc641c68e49ff6e11d6dceb

File: corehq/apps/users/static/users/js/mobile_workers.ng.js (hunk 4 above)
Comment (czue) [inline corehq/apps/users/static/users/js/mobile_workers.ng.js]: did you mean to leave this in? if so, should be `console.err` or include more information about the error?

**Comment by czue** [comment]
nice this looks good to me, thanks for the fast turnaround. not sure if you want someone more familiar with angular/js to take a pass but :+1: on :white_check_mark: from my side.

File: corehq/apps/users/static/users/js/mobile_workers.ng.js (hunk 4 above)
Comment (dimagimon) [inline corehq/apps/users/static/users/js/mobile_workers.ng.js]: Unexpected console statement. (no-console)

Comment (emord) [inline corehq/apps/users/static/users/js/mobile_workers.ng.js]: Yeah. I initially meant to leave this, but I agree that it doesn't really add anything. removed

**Pull request merged**